### ISSUE TITLE
Use cover cache

### ISSUE THREAD

**Issue by Ambrevar** [emacs-helm/helm-emms#12]
For a while EMMS has had support for the cover/artwork cache.
Among other niceties, it has the great benefit that it does not force users to name and size the covers in a particular way.

See `17.3 Displaying Covers` in the EMMS manual.

The `helm-emms-dired-transformer` hard-codes the path to covers though, which makes it impossible to use the cache.

A quick patch would be

```diff
(defun helm-emms-dired-transformer (candidates)
  (cl-loop with files
           for d in candidates
           ;; do (message "EMMS %S" (emms-browser-cache-thumbnail d 'small))
-           for cover = (pcase (expand-file-name "cover_small.jpg" d)
+           for cover = (pcase (or (emms-browser-cache-thumbnail d 'small)
+                                  (expand-file-name "cover_small.jpg" d) )
                         ((and c (pred file-exists-p)) c)
                         (_ (car emms-browser-default-covers)))
```

Better: make it a function so  that the user can customize how to retrieve the cover file.

Side question: on a large collection, this make `helm-emms` extremely slow to pop up everytime.  Would it be possible to cache/re-use the transformer?  Or maybe limit the computer covers to a 100 or something.

**Comment by thierryvolpiatto** [comment]
Pierre Neidhardt <notifications@github.com> writes:

> For a while EMMS has had support for the cover/artwork cache.
> Among other niceties, it has the great benefit that it does not force users to name and size the covers in a particular way.
>
> See 17.3 Displaying Covers in the EMMS manual.
>
> The helm-emms-dired-transformer hard-codes the path to covers though, which makes it impossible to use the cache.
>
> A quick patch would be
>
> (defun helm-emms-dired-transformer (candidates)
>   (cl-loop with files
>            for d in candidates
>            ;; do (message "EMMS %S" (emms-browser-cache-thumbnail d 'small))
> -           for cover = (pcase (expand-file-name "cover_small.jpg" d)
> +           for cover = (pcase (or (emms-browser-cache-thumbnail d 'small)
> +                                  (expand-file-name "cover_small.jpg" d) )
>                          ((and c (pred file-exists-p)) c)
>                          (_ (car emms-browser-default-covers)))
>
> Better: make it a function so that the user can customize how to retrieve the cover file.
>
> Side question: on a large collection, this make helm-emms extremely
> slow to pop up everytime.

Of course `emms-browser-cache-thumbnail` looks too slow to be used.

> Would it be possible to cache/re-use the transformer?

An alist with elements like (directory . cover_file) could be build at
startup, but helm-emms will be very slow at initial startup and what if
user modify or create a new music directory ? I think it is better to
let the user setup his music directories to fit to a fixed name like
cover_small.jpg or whatever.

-- 
Thierry

**Comment by Ambrevar** [comment]
> Of course `emms-browser-cache-thumbnail` looks too slow to be used.

I wrote the thumbnail caching code :)

Any idea how to speed it up?

**Comment by thierryvolpiatto** [comment]
Pierre Neidhardt <notifications@github.com> writes:

>> Of course `emms-browser-cache-thumbnail` looks too slow to be used.
>
> I wrote the thumbnail caching code :)
>
> Any idea how to speed it up?

There is no need to speed it up IMO.

You should assume the user already did the conversion of its cover files
(with your function or whatever) and stick to a fixed value for the
cover filename, IMO it is not the place to do (or check if done) the
conversion while listing the directories.

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.*


-- 
Thierry

**Comment by Ambrevar** [comment]
I think there is a misunderstanding.

I meant "speed up the cache queries".

With the cache, the conversion of the covers is already done and
`emms-browser-cache-thumbnail' will only retrieve the path of the cached
thumnails.

I'm not saying that Helm should do the thumbnail conversions: this is done
automatically, and only once.

Maybe a faster way would be to store the paths in a hash table.  But that might
belong to EMMS to do it.  I'll try to optimize the cache lookups, see what I can do.

**Comment by Ambrevar** [comment]
I've done some benchmarking and optimization.

Setup: modern desktop, 1500+ albums on a mechanical hard drive.

First of all, `helm-source-emms-dired` without covers takes 12s to complete.
This makes `helm-source-emms-dired` unusable on my machine (`helm-emms` is a
command that is supposed to be called frequently).

I've patch-optimized EMMS to speed up the thumbnail cache lookups.  It helps tremendously.

Benchmarks:

- `helm-source-emms-dired` with covers and unoptimized EMMS: forever...
- `helm-source-emms-dired` with covers and optimized EMMS: 25s
- `helm-source-emms-dired` without covers: 12s
- `(helm-walk-directory emms-source-file-default-directory)`: 6s.

`helm-walk-directory` alone is prohibitive for me.

So I tried the to replace it with the following:

``` lisp
(defun ambrevar/helm-walk-directory (dir)
  (split-string
   (with-temp-buffer
     (process-file "find" nil '(t nil) nil (file-truename dir) "-mindepth" "1" "-type" "d")
     (buffer-string))
   "\n"))
```

`ambrevar/helm-walk-directory` is almost instant but has some drawbacks:

- It's less portable (should run on all Unix-based systems though).
- It won't work on directories with linebreaks.
(There might be some workarounds though.)

Then I tried moving the candidate transformer to the
filtered-candidate-transformer (I added an unused _source parameter to
`helm-emms-dired-transformer`).

Resulting benchmark: <1.5s !!!

Questions:

- Is there a good reason covers are formatted in the candidate-transformer and
not in the filtered-candidate-transformer?

- What do you think of integrating my suggested tweaks as user options?

**Comment by thierryvolpiatto** [comment]
Pierre Neidhardt <notifications@github.com> writes:

> Questions:
>
> - Is there a good reason covers are formatted in the candidate-transformer and
> not in the filtered-candidate-transformer?

No particular reason, I installed a change to use it.

> - What do you think of integrating my suggested tweaks as user options?

I committed a change to use a customized fn to search recursively dirs,
you can provide a function using an external program (find) as a PR but
I want to keep `helm-emms-walk-directory` as default for portability,
you may want to modify as well the defcustom to use 
(choice (function :tag "..." fn).

Thanks.

-- 
Thierry

**Comment by Ambrevar** [comment]
Thanks for the changes!
See #13.
